starlette/applications.py

```diff
@@ -124,10 +132,15 @@ def mount(self, path: str, app: ASGIApp, name: str | None = None) -> None:
     def host(self, host: str, app: ASGIApp, name: str | None = None) -> None:
         self.router.host(host, app=app, name=name)  # pragma: no cover
 
-    def add_middleware(self, middleware_class: type, **options: typing.Any) -> None:
+    def add_middleware(
+        self,
+        middleware_class: typing.Callable[typing.Concatenate[ASGIApp, P], typing.Any],
+        *args: P.args,
```

Comment:
Now without the args. 👀

Reply:
This is quite unfortunate as ParamSpec **requires** both `args` and `kwargs` to be present in the signature. This is not a breaking change, but certainly redundant since the underlying API does not accept `args`.

I figured this is _good enough_ and a small price for better type annotations that guard against potential issues regarding missing arguments to a Middleware class. 

What's your take on this?

Reply:
Can we just make `args` work? if so I wouldn't mind just adding it.

Reply:
Should the API remain intact, the ideal type annotations would be something along these lines:

```python
OptionsT = TypeVar("OptionsT", bound=TypedDict)


class MiddlewareT(Protocol[OptionsT]):
    def __call__(self, app: ASGIApp, **options: Unpack[OptionsT]) -> Any:
        ...


class Middleware:
    def __init__(self, cls: MiddlewareT[OptionsT], **options: Unpack[OptionsT]) -> None:
        self.cls = cls
        self.options = options
```

hypothetically - as `Unpack` does not currently allow "higher order" generic arguments.

Using `(app: ASGIApp, **options: P.kwargs) -> Any` won't work as well. 

Until the type system supports such a use case, we have to accept trade-offs or change the API accordingly (e.g. by passing down `args` as @adriangb suggested).

Reply:
Passing down `args` could definitely work, although I'm not sure if this is a desired change. I guess that the `kwargs`-only approach was very much intentional as it forces the user to be explicit. What do you think?

Reply:
Maybe being explicit was more important when there was no type checking?

Reply:
Good point. I'll get `args` working and we'll continue the discussion from there.